serde-generate/src/ts.rs

```diff
@@ -0,0 +1,1079 @@
+            U64 => "Uint64LE".into(),
+            U128 => "BigNumber".into(),
+            F32 => "number".into(),
+            F64 => "number".into(),
+            Char => "string".into(),
+            Str => "string".into(),
+            Bytes => "Uint8Array".into(),
+
+            Option(format) => format!("Optional{}", self.quote_type(format)),
+            Seq(format) => format!("List{}", self.quote_type(format)),
+            Map { key, value } => format!("Map{}{}", self.quote_type(key), self.quote_type(value)),
+            Tuple(formats) => format!("Tuple{}", self.quote_types(formats)),
+            TupleArray {
+                content,
+                size: _size,
+            } => format!("ListTuple{}", self.quote_type(content),),
+            Variable(_) => panic!("unexpected value"),
+        }
+    }
+
+    fn enter_class(&mut self, name: &str, reserved_subclass_names: &[&str]) {
+        self.out.indent();
+        self.current_namespace.push(name.to_string());
+        for name in reserved_subclass_names {
+            let entry = self
+                .current_reserved_names
+                .entry(name.to_string())
+                .or_insert(0);
+            *entry += 1;
+        }
+    }
+
+    fn leave_class(&mut self, reserved_subclass_names: &[&str]) {
+        self.out.unindent();
+        self.current_namespace.pop();
+        for name in reserved_subclass_names {
+            let entry = self.current_reserved_names.get_mut(*name).unwrap();
+            *entry -= 1;
+            if *entry == 0 {
+                self.current_reserved_names.remove(*name);
+            }
+        }
+    }
+
+    fn quote_types(&self, formats: &[Format]) -> String {
+        formats
+            .iter()
+            .map(|f| self.quote_type(f))
+            .collect::<Vec<_>>()
+            .join("")
+    }
+
+    fn output_trait_helpers(&mut self, registry: &Registry) -> Result<()> {
+        let mut subtypes = BTreeMap::new();
+        for format in registry.values() {
+            format
+                .visit(&mut |f| {
+                    if Self::needs_helper(f) {
+                        subtypes.insert(Self::mangle_type(f), f.clone());
+                    }
+                    Ok(())
+                })
+                .unwrap();
+        }
+
+        let mut imported = Vec::new();
+        for subtype in subtypes.values() {
+            self.output_typedef_helper(subtype, &mut imported)?;
+        }
+
+        writeln!(self.out, "export class TraitHelpers {{")?;
+        let reserved_names = &[];
+        self.enter_class("TraitHelpers", reserved_names);
+        for (mangled_name, subtype) in &subtypes {
+            self.output_serialization_helper(mangled_name, subtype)?;
+            self.output_deserialization_helper(mangled_name, subtype)?;
+        }
+        self.leave_class(reserved_names);
+        writeln!(self.out, "}}\n")
+    }
+
+    fn mangle_type(format: &Format) -> String {
```

Comment:
This wouldn't work as the typing system of typescript is completely different. we can try to typedef it for all types, but it would end up pretty much the same, isn't it? or did I miss something?

Reply:
`common::mangle_type` is what you want to generate names indexed by a Serde type. (see other comment)